### ISSUE TITLE
Bucket name contains invalid characters

### ISSUE THREAD

**Issue by Es-Esa** [sqlitebrowser/dbhub.io#189]
Im trying to test my sqlite as live and get "**Bucket name contains invalid characters**"
What does this mean?

CREATE TABLE toimipisteet (
    id INTEGER PRIMARY KEY,
    nimi VARCHAR(255) NOT NULL,
    osoite VARCHAR(255) NOT NULL,
    puhelin VARCHAR(50)
);

CREATE TABLE palvelut (
    id INTEGER PRIMARY KEY,
    nimi VARCHAR(255) NOT NULL,
    kuvaus TEXT,
    hinta DECIMAL(10,2) NOT NULL,
    toimipiste_id INTEGER,
    FOREIGN KEY (toimipiste_id) REFERENCES toimipisteet(id)
);

CREATE TABLE vuokratilat (
    id INTEGER PRIMARY KEY,
    nimi VARCHAR(255) NOT NULL,
    kuvaus TEXT,
    hinta DECIMAL(10,2) NOT NULL,
    toimipiste_id INTEGER,
    FOREIGN KEY (toimipiste_id) REFERENCES toimipisteet(id)
);

CREATE TABLE asiakkaat (
    id INTEGER PRIMARY KEY,
    nimi VARCHAR(255) NOT NULL,
    osoite VARCHAR(255) NOT NULL,
    puhelin VARCHAR(50),
    email VARCHAR(255)
);

CREATE TABLE varaukset (
    id INTEGER PRIMARY KEY,
    toimipiste_id INTEGER,
    asiakas_id INTEGER,
    tila_id INTEGER,
    alku_pvm DATE NOT NULL,
    loppu_pvm DATE NOT NULL,
    FOREIGN KEY (toimipiste_id) REFERENCES toimipisteet(id),
    FOREIGN KEY (asiakas_id) REFERENCES asiakkaat(id),
    FOREIGN KEY (tila_id) REFERENCES vuokratilat(id)
);

CREATE TABLE laskut (
    id INTEGER PRIMARY KEY AUTOINCREMENT,
    varaus_id INTEGER,
    erapvm DATE NOT NULL,
    summa DECIMAL(10,2) NOT NULL,
    laskutustapa VARCHAR(10) NOT NULL,
    FOREIGN KEY (varaus_id) REFERENCES varaukset(id)
);

CREATE TABLE kayttajat (
    id INTEGER PRIMARY KEY,
    kayttajatunnus VARCHAR(255),
    salasana_hash VARCHAR(255),
    isadmin BOOLEAN,
    luotu DATETIME
);

CREATE TABLE varaukset_palvelut (
    varaus_id INTEGER,
    palvelu_id INTEGER,
    PRIMARY KEY (varaus_id, palvelu_id),
    FOREIGN KEY (varaus_id) REFERENCES varaukset(id),
    FOREIGN KEY (palvelu_id) REFERENCES palvelut(id)
);

**Comment by chrisjlocke** [comment]
How did you create the database?  How did you upload it?
I created a database using your SQL with no issues.
https://dbhub.io/chrisjlocke/db3.db

Can you give a screenshot of the error?

**Comment by Es-Esa** [comment]
I createid with dbBrowser 
![kuva](https://user-images.githubusercontent.com/74989195/232034541-0ebfed0d-7a67-4b2e-ba45-621ebdab1c78.png)

**Comment by chrisjlocke** [comment]
Is the filename you created in DB Browser for SQLite containing unicode characters?  Can you try using a more 'simple' name?
I'm not sure what that error refers to, so purely a guess.

**Comment by Es-Esa** [comment]
Try to use your db. Downloaded and Still saying the same :D

pe 14. huhtik. 2023 klo 14.48 Chris Locke ***@***.***>
kirjoitti:

**Comment by justinclift** [comment]
Oh wow.  That specific error is coming from the storage layer of our backend, and I've not seen it occur before.

What's the name of the database you're trying to use?

As a workaround for now, try using a database name that's made of just plain alphanumeric characters.  That *should* work, while we look into why it's not liking your existing file name.

**Comment by chrisjlocke** [comment]
> Try to use your db.

![image](https://user-images.githubusercontent.com/7872218/232041200-02e4b1cd-8927-44a5-a309-5cd2698c3fc5.png)

Did you click the green download button?  Its just called db3.db .. downloaded fine here...

**Comment by chrisjlocke** [comment]
Stupid question #2, what is your username in dbhub.io ?  Does _that_ have unicode characters?

**Comment by justinclift** [comment]
Oh, that might be it too.

That being said, we used to not allow unicode characters in usernames (eg plain alphanumeric was about it), and there's been an outstanding task (for ages) to re-visit that.

I didn't *think* that had been done yet though.

**Comment by justinclift** [comment]
As a Standard database, it seems to be ok (as expected):

https://dbhub.io/Es-Esa/db3.db

@chrisjlocke I'll take a look at this over the weekend, as it sounds like we might need to figure out a different approach for generating the object name to store in Minio.

Instead of basing it upon the database name directly, we could use a hash or randomly generated string instead and store that string in the PostgreSQL database backend.

**Comment by chrisjlocke** [comment]
I can download from that link OK...
At what point in the chain (if I can download the file OK) is @Es-Esa getting the error?  Its not caching it somewhere else is it?

**Comment by justinclift** [comment]
Nah.  That error message occurs at pretty much the final step of the upload process, where the file is transferred to Minio for storage.

For "standard" databases, as they're static snapshots we just calculate the sha256 checksum of the file and use that for the file name.  We can't do that with live databases though, so I just went with the simple "use the database name" for the object.

It sounds like the list of characters that Minio accepts is even narrow than what we do.  So, I think we'd better switch to using a generated object name for storing it in Minio.  That should completely avoid this problem.

It's pretty late here though (11pm), so I'll work on this tomorrow.  Hopefully @Es-Esa is ok just using "standard alphanumeric" characters in the database name in the meantime. :smile:

**Comment by justinclift** [comment]
As a data point, I have an initial working fix for the problem.  Going to get some sleep and review it tomorrow before pushing it to the live system though.

**Comment by justinclift** [comment]
The fix for this is now live on the production website.

@Es-Esa Any interest in trying again, to verify it's working from your point of view? :smile:

**Comment by Es-Esa** [comment]
Hei, im testing it right now, seems to work perfectly! thanks and sorry for not been around- I was out town xD

**Comment by justinclift** [comment]
Ahhh, no worries at all.  If you come across any other weirdness, please let us know.  We're fairly active developing this. :smile:
